grpcbp/server.go

```diff
@@ -0,0 +1,128 @@
+	// defined since the most recent time the number of outstanding RPCs became
+	// zero or the connection establishment. The current default value is
+	// infinity.
+	MaxConnectionIdle time.Duration `yaml:"maxConnectionIdle"`
+
+	// MaxConnectionAge is a duration for the maximum amount of time a
+	// connection may exist before it will be closed by sending a GoAway. A
+	// random jitter of +/-10% will be added to MaxConnectionAge to spread out
+	// connection storms. The current default value is infinity.
+	MaxConnectionAge time.Duration `yaml:"maxConnectionAge"`
+
+	// MaxConnectionAgeGrace is an additive period after MaxConnectionAge after
+	// which the connection will be forcibly closed. The current defualt value is
+	// infinity.
+	MaxConnectionAgeGrace time.Duration `yaml:"maxConnectionAgeGrace"`
+
+	// After a duration of this time if the server doesn't see any activity it
+	// pings the client to see if the transport is still alive. If set below 1s,
+	// a minimum value of 1s will be used instead. The current default value is 2
+	// hours.
+	Time time.Duration `yaml:"time"`
+
+	// After having pinged for keepalive check, the server waits for a duration
+	// of Timeout and if no activity is seen even after that the connection is
+	// closed. The current default value is 20 seconds.
+	Timeout time.Duration `yaml:"timeout"`
+
+	// MinTime is the minimum amount of time a client should wait before sending
+	// a keepalive ping.
+	MinTime time.Duration `yaml:"minTime"`
+
+	// If true, server allows keepalive pings even when there are no active
+	// streams(RPCs). If false, and client sends ping when there are no active
+	// streams, server will send GOAWAY and close the connection.
+	PermitWithoutStream bool `yaml:"permitWithoutStream"`
+
+	// RegisterServerFunc is used to pass in a generated gRPC service
+	// implementation and register it on the created gRPC.Server.
+	RegisterServerFunc func(*grpc.Server) `yaml:"-"`
+}
+
+// NewBaseplateServer returns a new gRPC implementation of a Baseplate server
+// with the given config.
+func NewBaseplateServer(bp baseplate.Baseplate, cfg ServerConfig) (baseplate.Server, error) {
+	lis, err := net.Listen("tcp", bp.GetConfig().Addr)
+	if err != nil {
+		return nil, err
+	}
+
+	kaep := keepalive.EnforcementPolicy{
+		MinTime:             cfg.MinTime,
+		PermitWithoutStream: cfg.PermitWithoutStream,
+	}
+
+	kasp := keepalive.ServerParameters{
+		MaxConnectionIdle:     cfg.MaxConnectionIdle,
+		MaxConnectionAge:      cfg.MaxConnectionAge,
+		MaxConnectionAgeGrace: cfg.MaxConnectionAgeGrace,
+		Time:                  cfg.Time,
+		Timeout:               cfg.Timeout,
+	}
+
+	middlewares := BaseplateDefaultMiddlewares(DefaultMiddlewaresArgs{
+		EdgeContextImpl: bp.EdgeContextImpl(),
+	})
+
+	server := grpc.NewServer(
+		grpc.KeepaliveEnforcementPolicy(kaep),
+		grpc.KeepaliveParams(kasp),
+		middlewares,
+	)
+	cfg.RegisterServerFunc(server)
```

Comment:
Hiding all of this away makes it really hard to use in tests.

I would have a top-level function that takes in your config and returns a grpc.Server

Reply:
Would this change the signature of the constructor though? Trying to figure out what type of testing we want to make more accessible. I'd think generally we want to be able to have users define their own server options altogether.